In [101]:
import pandas as pd

df = pd.read_csv('prueba2.csv', sep=';', encoding='utf-8')
df

id_sale  unit_price  quantity  total  id_track  \
0           1        0.99         1   0.99         2   
1           2        0.99         1   0.99         4   
2           3        0.99         1   0.99         6   
3           4        0.99         1   0.99         8   
4           5        0.99         1   0.99        10   
...       ...         ...       ...    ...       ...   
2124     2125        0.99         1   0.99      3020   
2125     2126        0.99         1   0.99      3029   
2126     2127        0.99         1   0.99      3038   
2127     2128        0.99         1   0.99      3047   
2128     2129        0.99         1   0.99      3056   

                                             track_name  milliseconds  \
0                                     Balls to the Wall        342562   
1                                     Restless and Wild        252051   
2                                 Put The Finger On You        205662   
3                                      Inject The Venom        210834   
4                                            Evil Walks        263497   
...                                                 ...           ...   
2124                           Um Homem Chamado Alfredo        151640   
2125                                   Looking For Love        391941   
2126                                    Sweet Lady Luck        273737   
2127  Feirinha da Pavuna/Luz do Repente/Bagaço da La...        107206   
2128                                   Samba pras moças        152816   

         bytes  id_album                            album_title  ...  \
0      5510424         2                      Balls to the Wall  ...   
1      4331779         3                      Restless and Wild  ...   
2      6713451         1  For Those About To Rock We Salute You  ...   
3      6852860         1  For Those About To Rock We Salute You  ...   
4      8611245         1  For Those About To Rock We Salute You  ...   
...        ...       ...                                    ...  ...   
2124   4976227       247                     Vinicius De Moraes  ...   
2125  12769847       141                          Greatest Hits  ...   
2126   8919163       141                          Greatest Hits  ...   
2127   3593684       248                       Ao Vivo [IMPORT]  ...   
2128   5121366       248                       Ao Vivo [IMPORT]  ...   

           employee_title employee_birthDate  employee_hireDate  \
0     Sales Support Agent         1965-03-03         2003-10-17   
1     Sales Support Agent         1965-03-03         2003-10-17   
2     Sales Support Agent         1947-09-19         2003-05-03   
3     Sales Support Agent         1947-09-19         2003-05-03   
4     Sales Support Agent         1947-09-19         2003-05-03   
...                   ...                ...                ...   
2124  Sales Support Agent         1973-08-29         2002-04-01   
2125  Sales Support Agent         1973-08-29         2002-04-01   
2126  Sales Support Agent         1973-08-29         2002-04-01   
2127  Sales Support Agent         1973-08-29         2002-04-01   
2128  Sales Support Agent         1973-08-29         2002-04-01   

       employee_adress  employee_city employee_country  id_invoice  \
0         7727B 41 Ave        Calgary           Canada           1   
1         7727B 41 Ave        Calgary           Canada           1   
2     683 10 Street SW        Calgary           Canada           2   
3     683 10 Street SW        Calgary           Canada           2   
4     683 10 Street SW        Calgary           Canada           2   
...                ...            ...              ...         ...   
2124     1111 6 Ave SW        Calgary           Canada         411   
2125     1111 6 Ave SW        Calgary           Canada         411   
2126     1111 6 Ave SW        Calgary           Canada         411   
2127     1111 6 Ave SW        Calgary           Canada         411   
2128     1111 6 

In [102]:
# listame la columna genre_name de la tabla df sin repetir valores
df['genre_name'].unique()


array(['Rock', 'Jazz', 'Metal', 'Alternative & Punk', 'Rock And Roll',
       'Blues', 'Latin', 'Reggae', 'Pop', 'Soundtrack', 'Bossa Nova',
       'Easy Listening', 'Heavy Metal', 'R&B/Soul', 'Electronica/Dance',
       'World', 'Hip Hop/Rap', 'Alternative', 'Classical'], dtype=object)

In [115]:

#Eliminando valores nulos
df = df.dropna()
df.shape

# Selecciona solo las columnas relevantes para el análisis de market basket
data_mba = df[['id_invoice', 'genre_name']]

# Elimina duplicados para tener solo una entrada por pedido y álbum
data_mba = data_mba.drop_duplicates()

# Crea una tabla pivote donde cada fila es un pedido, cada columna es un álbum y los valores son 1 si el género fue comprado en ese pedido, 0 de lo contrario
basket = pd.pivot_table(data_mba, index='id_invoice', columns='genre_name', aggfunc=lambda x: 1, fill_value=0)

basket

genre_name  Alternative  Alternative & Punk  Blues  Bossa Nova  Classical  \
id_invoice                                                                  
1                     0                   0      0           0          0   
2                     0                   0      0           0          0   
3                     0                   0      0           0          0   
4                     0                   0      0           0          0   
5                     0                   1      1           0          0   
...                 ...                 ...    ...         ...        ...   
407                   0                   0      0           0          0   
408                   0                   0      0           0          0   
409                   0                   0      0           0          0   
410                   0                   0      0           0          0   
411                   0                   0      0           0          0   

genre_name  Easy Listening  Electronica/Dance  Heavy Metal  Hip Hop/Rap  Jazz  \
id_invoice                                                                      
1                        0                  0            0            0     0   
2                        0                  0            0            0     0   
3                        0                  0            0            0     0   
4                        0                  0            0            0     1   
5                        0                  0            0            0     1   
...                    ...                ...          ...          ...   ...   
407                      0                  0            0            0     0   
408                      0                  0            0            0     0   
409                      0                  0            0            0     0   
410                      0                  0            0            0     0   
411                      0                  0            0            0     0   

genre_name  Latin  Metal  Pop  R&B/Soul  Reggae  Rock  Rock And Roll  \
id_invoice                                                             
1               0      0    0         0       0     1              0   
2               0      0    0         0       0     1              0   
3               0      0    0         0       0     1              0   
4               0      1    0         0       0     1              0   
5               1      1    0         0       0     0              1   
...           ...    ...  ...       ...     ...   ...            ...   
407             0      0    0         0       0     1              0   
408             0      0    0         0       0     1              0   
409             0      0    0         0       0     1              0   
410             0      0    0         0       0     1              0   
411             1      1    0         0       1     1              0   

genre_name  Soundtrack  World  
id_invoice                     
1                    0      0  
2                    0      0  
3                    0      0  
4                    0      0  
5                    0      0  
...                ...    ...  
407                  0      0  
408                  0      0  
409                  0      0  
410                  0      0  
411                  0      0  

[399 rows x 19 columns]

In [104]:
def encode_units(x):
    if x <= 0:
        return 0
    if x >= 1:
        return 1

basket_encode_plus = basket.applymap(encode_units)
basket_encode_plus

C:\Users\G1918543\AppData\Local\Temp\ipykernel_2776\1198150159.py:7: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  basket_encode_plus = basket.applymap(encode_units)


genre_name  Alternative  Alternative & Punk  Blues  Bossa Nova  Classical  \
id_invoice                                                                  
1                     0                   0      0           0          0   
2                     0                   0      0           0          0   
3                     0                   0      0           0          0   
4                     0                   0      0           0          0   
5                     0                   1      1           0          0   
...                 ...                 ...    ...         ...        ...   
407                   0                   0      0           0          0   
408                   0                   0      0           0          0   
409                   0                   0      0           0          0   
410                   0                   0      0           0          0   
411                   0                   0      0           0          0   

genre_name  Easy Listening  Electronica/Dance  Heavy Metal  Hip Hop/Rap  Jazz  \
id_invoice                                                                      
1                        0                  0            0            0     0   
2                        0                  0            0            0     0   
3                        0                  0            0            0     0   
4                        0                  0            0            0     1   
5                        0                  0            0            0     1   
...                    ...                ...          ...          ...   ...   
407                      0                  0            0            0     0   
408                      0                  0            0            0     0   
409                      0                  0            0            0     0   
410                      0                  0            0            0     0   
411                      0                  0            0            0     0   

genre_name  Latin  Metal  Pop  R&B/Soul  Reggae  Rock  Rock And Roll  \
id_invoice                                                             
1               0      0    0         0       0     1              0   
2               0      0    0         0       0     1              0   
3               0      0    0         0       0     1              0   
4               0      1    0         0       0     1              0   
5               1      1    0         0       0     0              1   
...           ...    ...  ...       ...     ...   ...            ...   
407             0      0    0         0       0     1              0   
408             0      0    0         0       0     1              0   
409             0      0    0         0       0     1              0   
410             0      0    0         0       0     1              0   
411             1      1    0         0       1     1              0   

genre_name  Soundtrack  World  
id_invoice                     
1                    0      0  
2                    0      0  
3                    0      0  
4                    0      0  
5                    0      0  
...                ...    ...  
407                  0      0  
408                  0      0  
409                  0      0  
410                  0      0  
411                  0      0  

[399 rows x 19 columns]

In [105]:
basket_filter_plus = basket_encode_plus[(basket_encode_plus > 0).sum(axis=1) >= 2]
basket_filter_plus

genre_name  Alternative  Alternative & Punk  Blues  Bossa Nova  Classical  \
id_invoice                                                                  
4                     0                   0      0           0          0   
5                     0                   1      1           0          0   
11                    0                   0      0           0          0   
12                    0                   0      0           0          0   
15                    0                   1      0           0          0   
...                 ...                 ...    ...         ...        ...   
397                   0                   1      1           0          0   
402                   0                   0      0           0          0   
403                   0                   1      0           0          0   
404                   0                   1      0           0          0   
411                   0                   0      0           0          0   

genre_name  Easy Listening  Electronica/Dance  Heavy Metal  Hip Hop/Rap  Jazz  \
id_invoice                                                                      
4                        0                  0            0            0     1   
5                        0                  0            0            0     1   
11                       0                  0            0            0     0   
12                       0                  0            0            0     0   
15                       0                  0            0            0     1   
...                    ...                ...          ...          ...   ...   
397                      0                  0            0            0     0   
402                      0                  0            0            0     0   
403                      0                  0            0            0     0   
404                      0                  0            0            0     0   
411                      0                  0            0            0     0   

genre_name  Latin  Metal  Pop  R&B/Soul  Reggae  Rock  Rock And Roll  \
id_invoice                                                             
4               0      1    0         0       0     1              0   
5               1      1    0         0       0     0              1   
11              1      0    0         0       1     0              0   
12              1      1    1         0       0     1              0   
15              0      0    0         0       0     0              0   
...           ...    ...  ...       ...     ...   ...            ...   
397             0      0    0         0       0     1              0   
402             1      0    0         0       0     1              0   
403             1      0    0         0       0     0              0   
404             0      0    0         0       0     1              0   
411             1      1    0         0       1     1              0   

genre_name  Soundtrack  World  
id_invoice                     
4                    0      0  
5                    0      0  
11                   0      0  
12                   1      0  
15                   0      0  
...                ...    ...  
397                  0      0  
402                  0      0  
403                  0      0  
404                  0      0  
411                  0      0  

[156 rows x 19 columns]

In [106]:
from mlxtend.frequent_patterns import apriori
frequent_itemsets_plus = apriori(basket_filter_plus, min_support=0.03,
                                 use_colnames=True).sort_values('support', ascending=False).reset_index(drop=True)

frequent_itemsets_plus['length'] = frequent_itemsets_plus['itemsets'].apply(lambda x: len(x))

frequent_itemsets_plus

c:\Users\G1918543\AppData\Local\Programs\Python\Python312\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                           itemsets  length
0   0.814103                             (Rock)       1
1   0.435897                            (Latin)       1
2   0.403846               (Alternative & Punk)       1
3   0.352564                            (Metal)       1
4   0.346154         (Alternative & Punk, Rock)       2
..       ...                                ...     ...
66  0.032051         (Latin, Soundtrack, Metal)       3
67  0.032051  (Alternative & Punk, Rock, Blues)       3
68  0.032051         (Rock, Heavy Metal, Metal)       3
69  0.032051  (Alternative & Punk, Jazz, Metal)       3
70  0.032051                      (Heavy Metal)       1

[71 rows x 3 columns]

In [107]:
frequent_itemsets_plus[ (frequent_itemsets_plus['length']  == 2) &
                        (frequent_itemsets_plus['support'] >= 0.03) ]

support                          itemsets  length
4   0.346154        (Alternative & Punk, Rock)       2
5   0.326923                     (Rock, Latin)       2
6   0.294872                     (Rock, Metal)       2
8   0.160256                    (Latin, Metal)       2
9   0.153846                      (Rock, Jazz)       2
11  0.128205       (Alternative & Punk, Latin)       2
13  0.108974                     (Latin, Jazz)       2
15  0.096154                     (Jazz, Metal)       2
16  0.083333        (Alternative & Punk, Jazz)       2
17  0.083333       (Alternative & Punk, Metal)       2
19  0.083333                     (Rock, Blues)       2
23  0.076923                (Rock, Soundtrack)       2
24  0.070513                    (Rock, Reggae)       2
29  0.064103                  (R&B/Soul, Rock)       2
32  0.057692                    (Metal, Blues)       2
33  0.057692               (Latin, Soundtrack)       2
35  0.051282       (Alternative & Punk, Blues)       2
36  0.051282                    (Latin, World)       2
38  0.051282                     (Rock, World)       2
39  0.051282                       (Rock, Pop)       2
41  0.044872                     (Jazz, Blues)       2
44  0.044872                    (Metal, World)       2
45  0.044872               (Rock, Hip Hop/Rap)       2
52  0.038462               (Soundtrack, Metal)       2
54  0.038462  (Alternative & Punk, Soundtrack)       2
55  0.038462                   (Latin, Reggae)       2
56  0.038462                 (R&B/Soul, Latin)       2
57  0.038462         (Rock, Electronica/Dance)       2
59  0.032051              (Heavy Metal, Metal)       2
62  0.032051                    (Latin, Blues)       2
63  0.032051               (Rock, Heavy Metal)       2

In [108]:
from mlxtend.frequent_patterns import association_rules

In [117]:
association_rules(frequent_itemsets_plus, metric='lift',
                  min_threshold=1).sort_values('lift', ascending=False).reset_index(drop=True).head(10).sort_values('lift', ascending=False)

antecedents           consequents  antecedent support  \
0               (World)  (Rock, Latin, Metal)            0.057692   
1  (Rock, Latin, Metal)               (World)            0.128205   
2        (Latin, Metal)         (Rock, World)            0.160256   
3         (Rock, World)        (Latin, Metal)            0.051282   
4        (Latin, Metal)               (World)            0.160256   
5               (World)        (Latin, Metal)            0.057692   
6         (Heavy Metal)         (Rock, Metal)            0.032051   
7         (Rock, Metal)         (Heavy Metal)            0.294872   
8          (Soundtrack)  (Rock, Latin, Metal)            0.076923   
9  (Rock, Latin, Metal)          (Soundtrack)            0.128205   

   consequent support   support  confidence      lift  leverage  conviction  \
0            0.128205  0.044872    0.777778  6.066667  0.037475    3.923077   
1            0.057692  0.044872    0.350000  6.066667  0.037475    1.449704   
2            0.051282  0.044872    0.280000  5.460000  0.036654    1.317664   
3            0.160256  0.044872    0.875000  5.460000  0.036654    6.717949   
4            0.057692  0.044872    0.280000  4.853333  0.035626    1.308761   
5            0.160256  0.044872    0.777778  4.853333  0.035626    3.778846   
6            0.294872  0.032051    1.000000  3.391304  0.022600         inf   
7            0.032051  0.032051    0.108696  3.391304  0.022600    1.085991   
8            0.128205  0.032051    0.416667  3.250000  0.022189    1.494505   
9            0.076923  0.032051    0.250000  3.250000  0.022189    1.230769   

   zhangs_metric  
0       0.886297  
1       0.957983  
2       0.972737  
3       0.861004  
4       0.945474  
5       0.842566  
6       0.728477  
7       1.000000  
8       0.750000  
9       0.794118

A partir de los resultados de las reglas de asociación anteriores, se puede ver que 'World' y 'Rock, Latin, Metal' son los elementos que tienen mayor asociación entre sí, ya que estos dos elementos tienen el valor de "elevación" mas alto. A mayor grado de elevación, mayor será la asociación entre los elementos. Al contar con un valor de elevación 6.06, siendo superior a 1, significa que los elementos tienen una asociación alta. Por lo que estos géneros son muy buenos para venderse juntos.


Adicionalmente, se puede ver que el valor de soporte de (World) y (Rock, Latin, Metal) es de 0.777778, lo que quiere decir que hay un 77% del total de transacciones en las que estos articulos se vendieron juntos

De la confianza se puede extraer que un cliente tenderá acomprar Heavy Metal antes de comprar Rock o Metal

In [113]:
rules = association_rules(frequent_itemsets_plus, metric='lift', min_threshold=1)
rules['antecedents_len'] = rules['antecedents'].apply(lambda x: len(x))
rules['consequents_len'] = rules['consequents'].apply(lambda x: len(x))

# Filtrar las reglas donde la longitud de antecedentes y consecuentes es 1
filtered_rules = rules[(rules['antecedents_len'] == 1) & (rules['consequents_len'] == 1)]

# Ordenar y mostrar las primeras 10 reglas
filtered_rules = filtered_rules.sort_values('lift', ascending=False).reset_index(drop=True).head(10)
filtered_rules

antecedents    consequents  antecedent support  consequent support  \
0        (Metal)  (Heavy Metal)            0.352564            0.032051   
1  (Heavy Metal)        (Metal)            0.032051            0.352564   
2        (World)        (Metal)            0.057692            0.352564   
3        (Metal)        (World)            0.352564            0.057692   
4        (World)        (Latin)            0.057692            0.435897   
5        (Latin)        (World)            0.435897            0.057692   
6        (Latin)   (Soundtrack)            0.435897            0.076923   
7   (Soundtrack)        (Latin)            0.076923            0.435897   
8        (Blues)         (Jazz)            0.134615            0.211538   
9         (Jazz)        (Blues)            0.211538            0.134615   

    support  confidence      lift  leverage  conviction  zhangs_metric  \
0  0.032051    0.090909  2.836364  0.020751    1.064744       1.000000   
1  0.032051    1.000000  2.836364  0.020751         inf       0.668874   
2  0.044872    0.777778  2.206061  0.024532    2.913462       0.580175   
3  0.044872    0.127273  2.206061  0.024532    1.079728       0.844413   
4  0.051282    0.888889  2.039216  0.026134    5.076923       0.540816   
5  0.051282    0.117647  2.039216  0.026134    1.067949       0.903409   
6  0.057692    0.132353  1.720588  0.024162    1.063885       0.742424   
7  0.057692    0.750000  1.720588  0.024162    2.256410       0.453704   
8  0.044872    0.333333  1.575758  0.016395    1.182692       0.422222   
9  0.044872    0.212121  1.575758  0.016395    1.098373       0.463415   

   antecedents_len  consequents_len  
0                1                1  
1                1                1  
2                1                1  
3                1                1  
4                1                1  
5                1                1  
6                1                1  
7                1                1  
8                1                1  
9                1                1

A partir de los resultados obtenidos de las reglas de asociación, se puede observar que los géneros musicales Metal y Heavy Metal presentan una fuerte correlación entre sí. Esta asociación se evidencia con un valor de "elevación" superior a los demás, específicamente un valor de 2.83, lo cual es considerablemente alto. Esto sugiere que estos dos géneros tienden a ser adquiridos conjuntamente, lo que los convierte en una excelente combinación para promociones o ventas cruzadas.

En cuanto a los géneros Latin y Soundtrack, el análisis revela un valor de soporte de 0.057. Esto significa que en el 5.7% del total de las transacciones, estos dos géneros se vendieron juntos. Aunque este porcentaje puede parecer pequeño, no debe ser ignorado ya que representa una oportunidad de venta conjunta.

El análisis de la confianza nos permite extraer información adicional interpretando el antecedente y el consecuente. En el caso de Heavy Metal y Metal, con un valor de confianza de 1, se puede inferir que un cliente que compra Heavy Metal probablemente también comprará Metal. Esta información es valiosa para la estrategia de marketing, ya que se podría ofrecer descuentos en el género Metal después de que un cliente haya comprado Heavy Metal, incentivando así la compra de ambos géneros.